### PR TITLE
Add ValidScopedCSSClass theme check

### PR DESCRIPTION
**Pull request by aswamy**
## 🗒️ Summary

Adds a new theme check `ValidScopedCSSClass` that warns when a CSS class used in an HTML `class` attribute may be defined outside the scope of the current file.

This helps theme developers catch cases where a CSS class is used in a file but is only defined in a stylesheet tag that isn't accessible from that file's scope — indicating a potential dependency issue or misplaced style.

## ❓ What is considered "in scope"

A CSS class is considered in scope for a file if it is defined in any of the following:

- **Local `{% stylesheet %}` tag** — the file's own stylesheet
- **Ancestor `{% stylesheet %}` tags** — any file that directly renders this file (via `{% render %}`, `{% section %}`, etc.), traversed upward through the full ancestor chain. Only `direct` type references count (not `indirect` or `preset`)
- **Rendered snippet stylesheets** — any snippet the file renders via `{% render %}`, including that snippet's own snippet descendants (recursive)
- **Ancestor-rendered snippet stylesheets** — any snippet rendered by an ancestor file, including descendants
- **CSS asset files** — all `.css` files in the theme's `assets/` directory are globally in scope

## 👍 What is NOT reported

- **Classes not defined anywhere in the theme** — classes from external sources (CDNs, utility frameworks like Tailwind, inline JS injection, etc.) are silently ignored. The check only reports a class if it exists somewhere in the theme but isn't reachable from the current file's scope
- **Dynamic class attributes** — Liquid output tags (`{{ variable }}`) in class attributes are skipped since they can't be statically analyzed

#### Via VSCode Extension
- Checkout this branch
- Run `yarn install && yarn build`
- Press F5 to run the VScode extension in another window (might be `fn + F5` for some)
- Open a theme (or create one using Shopify CLI by running `npx shopify theme init`)
- Add a few CSS classes in stylesheet tags or `assets/*.css` files

#### Run cli-like environment
- Run the following command against the theme from this repo's root `node packages/theme-check-node/dist/cli.js [path-to-theme] > result` (replace [path-to-theme] with an actual path to theme)
  - See the theme check results from above appear in the `result` file it generated

### CODE REVIEW

File: packages/theme-graph/package.json [hunk 1]
```diff
@@ -31,9 +31,11 @@
   "dependencies": {
     "@shopify/liquid-html-parser": "^2.9.2",
     "@shopify/theme-check-common": "^3.24.0",
-    "@shopify/theme-check-node": "^3.24.0",
     "acorn": "^8.14.1",
     "acorn-walk": "^8.3.4",
     "vscode-uri": "^3.0.7"
+  },
+  "devDependencies": {
+    "@shopify/theme-check-node": "^3.24.0"
```
Comment (aswamy) [inline packages/theme-graph/package.json:39]: I had to move it to a dev-dependency so we can let theme-check-node be dependent on the theme-graph.

Why? Because CLI needs the theme-graph to be able to run the theme-check to detect cross-file CSS usage.

File: packages/theme-check-node/src/index.ts [hunk 2]
```diff
@@ -135,6 +137,22 @@ export async function themeCheckRun(
...
+  const getSectionSchemaFn = async (name: string) => sectionSchemas.get(name)?.();
+  const getBlockSchemaFn = async (name: string) => blockSchemas.get(name)?.();
+
+  let themeGraph: ThemeGraph | undefined;
+  try {
+    themeGraph = await buildThemeGraph(config.rootUri, {
+      fs: NodeFileSystem,
+      getSectionSchema: getSectionSchemaFn as any,
+      getBlockSchema: getBlockSchemaFn as any,
+      getWebComponentDefinitionReference: () => undefined,
```
Comment (aswamy) [inline packages/theme-check-node/src/index.ts:173]: I don't want to recreate the whole graph because we don't need it for this theme-check, but might silently error out if we add a theme-check that needs this in the future 🤔 

(but then again, we would silently fail if `getReferences` or `getDependencies` wasn't defined either so maybe not a HUGE deal 🤷)

File: packages/theme-language-server-common/src/server/startServer.ts [hunk 3]
```diff
@@ -459,6 +459,9 @@ export function startServer(
         {
           globPattern: '**/*.liquid',
         },
+        {
+          globPattern: '**/assets/*.css',
```
Comment (aswamy) [inline packages/theme-language-server-common/src/server/startServer.ts:463]: Need this or CSS files in assets are going to remain cached from initial load when running themecheck

File: packages/theme-check-common/src/checks/valid-scoped-css-class/index.ts
Comment (graygilmore) [inline packages/theme-check-common/src/checks/valid-scoped-css-class/index.ts:1]: I tested a hilarious amount of edge cases and I think I found a few!

**Case sensitivity**

As far as I know HTML attributes aren't case sensitive so these should warn but currently don't:

```html
<div CLASS="out-of-scope-class"></div>
<div Class="out-of-scope-class"></div>
```

**SVGs**

```html
<svg class="out-of-scope-class">
  <rect class="out-of-scope-class" />
  <g class="out-of-scope-class"></g>
  <circle class="out-of-scope-class" />
  <path class="out-of-scope-class" />
</svg>
```

**Conditional classes**

I don't think we can actually do anything about these since we can't know the value of a dynamic class but I thought I'd show them just in case:

```liquid
<div class="scoped {% if true %}out-of-scope-class{% endif %}"></div>
```

`scoped` is checked correctly but inside the conditional isn't.

**Dynamic classes**

Similar to the above I know we can't validate dynamic classes but I wonder if this should still warn:

```liquid
{% assign butts = "butts" %}
<div class="out-of-scope-class{{ butts }} class1 class2"></div>
```

I think Claude found a suggestion for this one.

**Orphaned snippets**

I don't know if we actually want to cover this case but if you have two orphaned snippets that create their own little tree you'll get a false positive warning when running theme check (but you won't get the warning when viewing the file in the editor it seems?):

```liquid
<!-- snippets/orphan.liquid -->
{% render 'orphan-styles' %}
<div class="orphan-button">Click me</div>
```

```liquid
<!-- snippets/orphan-styles.liquid -->
{% stylesheet %}
  .orphan-button {
    color: red;
  }
{% endstylesheet %}
```

File: packages/theme-check-node/src/index.ts [hunk 4]
```diff
@@ -134,6 +138,43 @@ export async function themeCheckRun(
...
+  // Build theme graph for cross-file checks (e.g. ValidScopedCSSClass)
+  const getSectionSchemaFn = async (name: string) => sectionSchemas.get(name)?.();
+  const getBlockSchemaFn = async (name: string) => blockSchemas.get(name)?.();
+
+  let themeGraph: ThemeGraph | undefined;
+  try {
+    themeGraph = await buildThemeGraph(config.rootUri, {
+      fs: NodeFileSystem,
+      getSectionSchema: getSectionSchemaFn as any,
+      getBlockSchema: getBlockSchemaFn as any,
```
Comment (graygilmore) [inline packages/theme-check-node/src/index.ts]: Are we able to avoid the `as any`?

File: packages/theme-graph/package.json (hunk 1 above)
Comment (graygilmore) [reply packages/theme-graph/package.json:39]: I think this will blow up the theme-graph binary: https://github.com/Shopify/theme-tools/blob/main/packages/theme-graph/bin/theme-graph#L10

File: packages/theme-check-common/src/checks/valid-scoped-css-class/index.ts [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    schema: {},
+    targets: [],
+  },
+
+  create(context) {
+    const localCSSClasses = new Set<string>();
+    const usedClasses: { className: string; startIndex: number; endIndex: number }[] = [];
+
+    return {
+      async LiquidRawTag(node: LiquidRawTag) {
```
Comment (graygilmore) [inline packages/theme-check-common/src/checks/valid-scoped-css-class/index.ts:93]: 🎨 **Code Style**: The `LiquidRawTag`, `HtmlElement`, `HtmlVoidElement`, and `HtmlSelfClosingElement` visitors are declared `async` but perform only synchronous operations — Set/Array mutations with no `await`. Other checks in the codebase use non-async visitors for synchronous work. While functionally harmless, this adds unnecessary microtask overhead per node visit.

**Suggestion:** Remove `async` from the synchronous visitors. The `onCodePathEnd` visitor correctly remains `async` since it uses `await`.

```suggestion
      LiquidRawTag(node: LiquidRawTag) {
```

File: packages/theme-check-common/src/checks/valid-scoped-css-class/index.ts [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+        // 6. Extract CSS classes from all reachable snippet stylesheet tags
+        const snippetClassResults = await Promise.all(
+          allSnippetUris.map((uri) => getCSSClassesForURI(uri)),
+        );
+        for (const classes of snippetClassResults) {
+          for (const cls of classes) inScopeClasses.add(cls);
+        }
+
+        // 7. Collect ALL CSS classes defined anywhere in the theme
+        const allThemeClasses = await getAllThemeCSSClasses(fs, toUri, getCSSClassesForURI);
```
Comment (graygilmore) [inline packages/theme-check-common/src/checks/valid-scoped-css-class/index.ts:173]: ⚡ **Performance**: `getAllThemeCSSClasses` is called inside `onCodePathEnd`, which runs once per checked file. Each invocation does a full recursive directory traversal and reads+parses every `.liquid` and `.css` file in the theme. For a theme with N files using class attributes and M total theme files, this means N full-theme scans. Even in the CLI path where `getCSSClassesForURI` is memoized per-URI, the directory listing itself repeats on every invocation. The same applies to `getAssetCSSClasses`.

The set of all theme CSS classes is a theme-wide constant. The established pattern in this codebase (e.g., `sectionSchemas`, `blockSchemas`, `docDefinitions`) is to precompute expensive run-scoped data once and share it through injected dependencies. Since this check is enabled by default, the cost scales with every theme that uses class attributes.

**Suggestion:** Cache the theme-wide class index once per lint run and share it through the injected dependencies, following the pattern used by other expensive run-scoped lookups in the codebase. For example, add `getAllThemeCSSClasses?: () => Promise<Set<string>>` to the `Dependencies` interface and have the host compute and cache it once. Alternatively, memoize the result within the check's `create()` closure.

File: packages/theme-check-common/src/utils/styles.ts [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+
+/** Extract used class names from an HTML element's class attribute. */
+export function collectUsedClasses(
+  attributes: AttributeNode[],
+  usedClasses: { className: string; startIndex: number; endIndex: number }[],
+): void {
+  for (const attr of attributes) {
+    if (!isValuedHtmlAttribute(attr) || !isAttr(attr, 'class')) continue;
+
+    for (const valueNode of attr.value) {
```
Comment (graygilmore) [inline packages/theme-check-common/src/utils/styles.ts]: 🐛 **Bug**: The class extraction helper tokenizes each `TextNode` independently, but the parser represents a class attribute as alternating text and Liquid nodes. A value like `class="btn-{{ size }}"` contributes `btn-` as if it were a complete static class. This is inconsistent with the intended behavior of skipping dynamic class usage — the check correctly skips fully dynamic classes but still emits partial fragments around Liquid interpolation, which can produce false positives for common dynamic-class patterns.

**Suggestion:** Consider accumulating class tokens across the full attribute value and discarding the current token whenever a non-text Liquid node interrupts it. That keeps `class="static {{ dynamic }}"` valid for `static` while avoiding false positives for partial tokens like `btn-` from `btn-{{ size }}`.

File: packages/theme-check-common/src/checks/valid-scoped-css-class/index.ts (hunk 5 above)
Comment (aswamy) [reply packages/theme-check-common/src/checks/valid-scoped-css-class/index.ts:93]: We can't do this because it breaks the type of the object being returned - each method needs to return `Promise<void>` - we could just do that for each method, but that's the same overhead of adding `async` to methods 🤷

**Review by stephanie-shopify — COMMENTED** [review]
Ran a top hat, found one possible case to look at
Grandparent chain:
 Layout → section → snippet: When a layout file defines a CSS class and renders a section, which in turn renders a snippet that uses that class, the check warns that the class is out of scope.                                                                 
                                                                                                                                                                                                                                                                             
 The setup:                                                                                                                                                                                                                                                                  
 - layout/theme.liquid has `{% stylesheet %} .layout-class { ... } {% endstylesheet %}` and `{{ content_for_layout }}`                                                                                                                                                           
 - sections/section.liquid renders ` {% render 'deep-snippet' %} `                                                                                                                                                                                                              
 - snippets/deep-snippet.liquid uses `<div class="layout-class"> `

**Comment by aswamy** [comment]
> Ran a top hat, found one possible case to look at Grandparent chain: Layout → section → snippet: When a layout file defines a CSS class and renders a section, which in turn renders a snippet that uses that class, the check warns that the class is out of scope.
> 
> The setup:
> 
> * layout/theme.liquid has `{% stylesheet %} .layout-class { ... } {% endstylesheet %}` and `{{ content_for_layout }}`
> * sections/section.liquid renders `{% render 'deep-snippet' %}`
> * snippets/deep-snippet.liquid uses `<div class="layout-class"> `

@stephanie-shopify Layout's are not considered *direct* parents of sections. Their relation is sort of just implied because of `content_for_layout`.

You'll see in the side bar, that the parent of `sections/hello-world.liquid` isn't a layout file, but the template json. I assume that's true for your situation as well. There are no _true_ parents for those json files.

image

File: packages/theme-check-common/src/checks/valid-scoped-css-class/index.ts (hunk 6 above)
Comment (aswamy) [reply packages/theme-check-common/src/checks/valid-scoped-css-class/index.ts:173]: The issue isn't really a problem on VSCode Extension since we cache the entire fs. But on CLI, there is partial concern to what you're saying because even though `getCSSClassesForURI` is memoed, we are iterating over the files individually using `recursiveReadDirectory`. We could have just made `getAllThemeCSSClasses` like you suggested which would do it in bulk, but i feel like this breaks the current pattern we are going for [here](https://github.com/Shopify/theme-tools/blob/f5b8d0712d7cdf9d7e96acfb1be3e2c1d8ee8b3f/packages/theme-check-common/src/types.ts#L347-L395). We always pass a URI for a specific file and it returns its memoized content.

Ill try it out in a subsequent PR though because i still think we take around 18sec of wall-clock time and any gains is nice.

File: packages/theme-graph/package.json (hunk 1 above)
Comment (aswamy) [reply packages/theme-graph/package.json:39]: Im going to use `theme-check-common` instead of `theme-check-node` in the binary and move any functions over that doesn't exist. I dont think anyone uses this raw binary anyway.

**Review by stephanie-shopify — APPROVED** [review]

**Comment by aswamy** [comment]
@graygilmore Thanks for the thorough tophat and review Gray!

I've completed the following fixes:
✅ Case sensitivity
✅ SVGs - a little tougher than i thought because we consider the SVG tag a "raw" node so i had to add a helper to parse it
✅ Dynamic Classes - we do make sure we distinguish between `btn-{{ size }}` and `btn {{ size }}`. The first one the entire class is considered dynamic and discarded. The second one, we only check `btn` to see if it appears in other stylesheet tag


I think "Conditional classes" and "Orphaned snippets" might be a little bit too niche. We can always update it once we get feedback on this.

> if you have two orphaned snippets that create their own little tree you'll get a false positive warning when running theme check (but you won't get the warning when viewing the file in the editor it seems?):

That's a weird bug. I noticed that too. Happens in CLI but not VSCode extension? I wonder if it's a theme-graph issue for orphaned files because the orphan-styles file doesn't have any references in the side panel on VSCode. It should at least have `orphan`. Either way, orphaned files isn't too much of a concern for us.

File: packages/theme-check-common/src/utils/styles.ts [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+
+      if (tokens.length === 0) continue;
+
+      // Drop first token if it touches the preceding Liquid node (no leading whitespace)
+      if (prevIsLiquid && !value.startsWith(' ') && !value.startsWith('\n')) {
+        tokens.shift();
+      }
+
+      // Drop last token if it touches the following Liquid node (no trailing whitespace)
+      if (nextIsLiquid && !value.endsWith(' ') && !value.endsWith('\n')) {
```
Comment (graygilmore) [inline packages/theme-check-common/src/utils/styles.ts]: Last edge case! These two checks only check for new lines and spaces but HTML spec considers more than those as whitespace (tabs for example). Might want to use a regex to match all ASCII whitespace (I haven't validated the below):

```js
// startswith
if (prevIsLiquid && !/^\s/.test(value)) {}

// endswith
if (nextIsLiquid && !/\s$/.test(value)) {}
```

**Review by graygilmore — APPROVED** [review]
One more small edge case
